src/blas/backends/cublas/cublas_scope_handle.cpp

```diff
@@ -42,13 +42,13 @@ CublasScopedContextHandler::CublasScopedContextHandler(sycl::queue queue, sycl::
         : ih(ih),
           needToRecover_(false) {
     placedContext_ = new sycl::context(queue.get_context());
-    auto device = queue.get_device();
-    auto desired = sycl::get_native<sycl::backend::ext_oneapi_cuda>(*placedContext_);
+    auto cudaDevice = ih.get_native_device<sycl::backend::ext_oneapi_cuda>();
     CUresult err;
+    CUcontext desired;
     CUDA_ERROR_FUNC(cuCtxGetCurrent, err, &original_);
+    // Getting the primary context also sets it as the active context
+    CUDA_ERROR_FUNC(cuDevicePrimaryCtxRetain, err, &desired, cudaDevice);
```

Comment:
Should we expect a performance cost from this change? From what I understand `cuCtxSetCurrent` was expected to be called only once before, assuming the context active was not changed outside of oneMKL.
This constructor is called once before each calls to blas functions so I am wary that the cost may add up.

Reply:
The cost of `cuDevicePrimaryContextRetain` is minimal once the primary context is not being initialized for the first time, which it should not be here. A simple benchmark like this:
```c++
for (int i = 0; i < NUM_ITERATIONS; i++) {
    res = cuDevicePrimaryCtxRetain(&context, device);
    if (res != CUDA_SUCCESS) {
      printf("cuDevicePrimaryCtxRetain failed: res = %d\n", res);
      return -1;
    }

    res = cuDevicePrimaryCtxRelease(device);
    if (res != CUDA_SUCCESS) {
      printf("cuDevicePrimaryCtxRelease failed: res = %d\n", res);
      return -1;
    }
  }
```

Gives 32ns per loop, which calls `cuDevicePrimaryCtxRetain` and `cuDevicePrimaryCtxRelease`, once the context is already set and is not changing. So calls to this func are almost free.